L73-java-binderchannel/wireformat.md

```diff
@@ -0,0 +1,200 @@
+# The wire format of BinderTransport.
+
+All communication occurs via one-way binder
+[transactions](https://developer.android.com/reference/android/os/Binder#transact\(int,%20android.os.Parcel,%20android.os.Parcel,%20int\))
+which are made up of an integer “code”, and a [Parcel] containing arbitrary
+data.
+
+Three types of binder are involved in our implementation.
+
+*   The _endpoint binder_, owned by the BinderServer or OnDeviceServerEndpoint
+    instance, and returned from the host Android service's onBind method.
+*   The _client binder_, owned by the ClientTransport implementation within the
+    Channel.
+*   The _server binder_, owned by the ServerTransport implementation.
+
+A single endpoint binder exists for each BinderServer or OnDeviceServerEndpoint.
+A client binder / server binder pair is created for each transport pair.
+
+## Transaction Types
+
+There are two types of transactions, control and stream.
+
+### Transport-level Control Transactions
+
+For general transport lifecycle, using one of 1000 reserved transaction codes.
+
+*   Setup Transport
+    *   Transaction code: BinderTransport.SETUP_TRANSPORT (1)
+    *   From ClientTransport to host binder, and from ServerTransport to client
+        binder.
+*   Shutdown Transport
+    *   Transaction code: BinderTransport.SHUTDOWN_TRANSPORT (2)
+    *   From ClientTransport to server binder, or ServerTransport to client
+        binder.
+*   Acknowledge bytes.
+    *   Transaction code: BinderTransport.ACKNOWLEDGE_BYTES (3)
+    *   From either transport to its peer's binder. Reports reception of
+        transaction data for flow control.
+*   Ping
+    *   Transaction code: BinderTransport.PING (4)
+    *   Currently only sent from ClientTransport to server binder.
+*   Ping Response:
+    *   Transaction code: BinderTransport.PING_RESPONSE (5)
+    *   From SeverTransport to client binder or ClientTransport to server binder,
+        in response to a ping message.
+
+If either side of a transport receives an unrecognized control transacton,
+it will shutdown the transport gracefully.
+
+#### Transaction Data
+
+```
+major version = int;
+protocol extension flags = int;
+num bytes = int;
+ping id = int;
+
+setup transport transaction = major version, binder, [protocol extension flags];
+shutdown transport transaction = nothing;
+acknowledge bytes transaction = num bytes;
+ping transaction = ping id;
+ping response transaction = ping id;
+```
+
+During transport setup, if a server receves a setup transaction with a higher
```

Comment:
"higher protocol version" may easily be misinterpreted. I know you mean a version higher than it supports. And actually we don't say what should happen if the version is lower than the max supported by the server.

Not perfect, but maybe good enough:
> `setup transport transaction` is guaranteed to be backward-compatible and so is safe to process even if the version number is unknown to an implementation. During transport setup, the version sent by the client should be the largest version it supports. The server should respond with the actual version to be used, which will be equal to or less than the client-advertised version. The server must not fail if the client-advertised version is higher than supported. The server may fail if the client-advertised version is too low to be supported. The client may fail if the server-selected version is too low to be supported.

(I said "fail" there. It may be good to be more specific that the server would respond with SHUTDOWN_TRANSPORT and the client would simply release the binder and notify the client.)

s/receves/receives/

Reply:
You've written these things before. 8) Copied shamelessly with the SHUTDOWN_TRANSPORT detail.